src/models/addressBook/index.js

```diff
@@ -0,0 +1,180 @@
+/**
+ * Copyright 2020 Inrupt Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to use,
+ * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
+ * Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+import {
+  addStringNoLocale,
+  addUrl,
+  createSolidDataset,
+  createThing,
+  getSolidDataset,
+  getSourceUrl,
+  getThing,
+  getUrl,
+  saveSolidDatasetAt,
+  setThing,
+} from "@inrupt/solid-client";
+import { acl, dc, foaf, rdf, schema, vcard } from "rdf-namespaces";
+import { joinPath } from "../../stringHelpers";
+import { chain } from "../../solidClientHelpers/utils";
+import { ERROR_CODES, isHTTPError } from "../../error";
+
+/**
+ * @typedef AddressBook
+ * @type {object}
+ * @property {string} containerIri - The address to the container
+ * @property {object} dataset - The dataset the main index is stored in
+ * @property {object} thing - The thing that represents the main index itself
+ */
+
+/* Model constants */
+const CONTACTS_CONTAINER = "contacts/";
+
+export function vcardExtras(property) {
+  return `http://www.w3.org/2006/vcard/ns#${property}`;
+}
+
+export const NAME_EMAIL_INDEX_PREDICATE = vcardExtras("nameEmailIndex");
+export const NAME_GROUP_INDEX_PREDICATE = vcardExtras("groupIndex");
+export const VCARD_WEBID_PREDICATE = vcardExtras("WebId");
+export const INDEX_FILE = "index.ttl";
+export const PEOPLE_INDEX_FILE = "people.ttl";
+export const GROUPS_INDEX_FILE = "groups.ttl";
+export const PERSON_CONTAINER = "Person";
+export const GROUP_CONTAINER = "Group";
+
+const person = {
```

Comment:
This model is still referring to "person" and "group". Per our conversations, these concepts should not exist in this file and should be lazily created when a person or group is added to an address book for the first time.

Reply:
Ok, think I've been able to move out all "person" and "group" related info from AddressBook now - everything is now handled by the Contact model, including a method that updates the AddressBook main index when a group is created and group index doesn't exist from before.

Do note that this violates the constraints described in https://github.com/inrupt/pod-browser/blob/master/shapes/contacts/address-book.shex, which states that `vcard:nameEmailIndex` and `vcard:groupIndex` should be set. That said, I do like the lazy-creation approach, so if anything we can discuss with the SolidOS/Data Browser team whether we should allow that flexibility.